firedrake/extrusion_utils.py

```diff
@@ -150,17 +149,29 @@ def make_extruded_coords(extruded_topology, base_coords, ext_coords,
     else:
         raise NotImplementedError('Unsupported extrusion type "%s"' % extrusion_type)
 
-    # Dat to hold layer number
-    import firedrake.functionspace as fs
-    layer_fs = fs.FunctionSpace(extruded_topology, 'DG', 0)
-    layers = extruded_topology.layers
-    layer = op2.Dat(layer_fs.dof_dset,
-                    np.repeat(np.arange(layers-1, dtype=np.int32),
-                              extruded_topology.cell_set.total_size).reshape(layers-1, extruded_topology.cell_set.total_size).T.ravel(), dtype=np.int32)
     height = op2.Global(1, layer_height, dtype=float)
     op2.par_loop(kernel,
                  ext_coords.cell_set,
                  base_coords.dat(op2.READ, base_coords.cell_node_map()),
                  ext_coords.dat(op2.WRITE, ext_coords.cell_node_map()),
-                 layer(op2.READ, layer_fs.cell_node_map()),
-                 height(op2.READ))
+                 height(op2.READ),
+                 pass_layer_arg=True)
+
+
+def flat_entity_dofs(entity_dofs):
+    flat_entity_dofs = {}
+    for b, v in entity_dofs:
+        # v in [0, 1].  Only look at the ones, then grab the data from zeros.
+        if v == 0:
+            continue
+        flat_entity_dofs[b] = {}
+        for i in entity_dofs[(b, v)]:
+            # This line is fairly magic.
+            # It works because an interval has two points.
+            # We pick up the DoFs from the bottom point,
+            # then the DoFs from the interior of the interval,
+            # then finally the DoFs from the top point.
+            flat_entity_dofs[b][i] = (entity_dofs[(b, 0)][2*i] +
+                                      entity_dofs[(b, 1)][i] +
+                                      entity_dofs[(b, 0)][2*i+1])
+    return flat_entity_dofs
```

Comment:
We did, that's why f7c48b7848 removes the usage and turns it into a function.  Unless I missed other usages.